driver/src/orderbook/filtered_orderbook.rs

```diff
@@ -0,0 +1,160 @@
+use super::*;
+
+use serde::Deserialize;
+use std::collections::{HashMap, HashSet};
+use web3::types::{H160, U256};
+
+/// Data structure to specify what type of orders to filter
+#[derive(Debug, Default, Deserialize, PartialEq, Eq)]
+pub struct OrderbookFilter {
+    /// The token ids that should be filtered/
+    tokens: HashSet<u16>,
+
+    /// User addresses mapped to which of their orders to filter
+    users: HashMap<H160, UserOrderFilter>,
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq)]
+enum UserOrderFilter {
+    All,
+    OrderIds(HashSet<u16>),
+}
+
+pub struct FilteredOrderbookReader<'a> {
+    orderbook: &'a dyn StableXOrderBookReading,
+    filter: OrderbookFilter,
+}
+
+impl<'a> FilteredOrderbookReader<'a> {
+    pub fn new(orderbook: &'a dyn StableXOrderBookReading, filter: OrderbookFilter) -> Self {
+        Self { orderbook, filter }
+    }
+}
+
+impl<'a> StableXOrderBookReading for FilteredOrderbookReader<'a> {
+    fn get_auction_index(&self) -> Result<U256> {
+        self.orderbook.get_auction_index()
+    }
+
+    fn get_auction_data(&self, index: U256) -> Result<(AccountState, Vec<Order>)> {
+        let (state, orders) = self.orderbook.get_auction_data(index)?;
+        let filtered = orders
+            .into_iter()
+            .filter(|o| {
+                let user_filter = if let Some(user_filter) = self.filter.users.get(&o.account_id) {
+                    match user_filter {
+                        UserOrderFilter::All => true,
+                        UserOrderFilter::OrderIds(ids) => ids.contains(
+                            &o.batch_information
+                                .as_ref()
+                                .expect("StableX orders have batch information")
+                                .slot_index,
+                        ),
+                    }
+                } else {
+                    false
+                };
+                !self.filter.tokens.contains(&o.buy_token)
+                    && !self.filter.tokens.contains(&o.sell_token)
+                    && !user_filter
+            })
+            .collect();
+        Ok((state, filtered))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use dfusion_core::models::order::test_util::create_order_for_test;
+    use std::str::FromStr;
+
+    #[test]
+    fn test_filter_deserialization() {
+        let json = "{
```

Comment:
One minor suggestion I just though of, you can use a raw string literal here, which would make you not need any escape sequences:
```rust
let json = r#"{
  "foo": "bar"
}"#;
```